diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6a9a694..f8fb6e4 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -21,6 +21,7 @@ option(ASSERT2 "additional assertions" OFF)
 option(DEBUG "add debugging support" OFF)
 option(GPROF "add gprof support" OFF)
 option(GDB "add gdb support" OFF)
+option(NORMALIZED "normalize for A/B testing with Rust" OFF)
 option(GKRAND "enable GKRAND support" OFF)
 option(GKREGEX "enable GKREGEX support" OFF)
 option(OPENMP "enable OpenMP support" OFF)
@@ -73,7 +74,8 @@ target_include_directories(${PROJECT_NAME}
          $<INSTALL_INTERFACE:include>)
 
 target_link_libraries(${PROJECT_NAME}
-  PUBLIC $<$<NOT:$<C_COMPILER_ID:MSVC>>:m>)
+  PUBLIC $<$<NOT:$<C_COMPILER_ID:MSVC>>:m>
+  PUBLIC functions)
 
 set_target_properties(${PROJECT_NAME} PROPERTIES
   SOVERSION ${PROJECT_VERSION_MAJOR}
@@ -89,6 +91,7 @@ target_compile_definitions(${PROJECT_NAME}
          $<$<NOT:$<BOOL:${ASSERT2}>>:NDEBUG2>
          $<$<BOOL:${DEBUG}>:DEBUG>
          $<$<BOOL:${GKRAND}>:GKRAND>
+         $<$<BOOL:${NORMALIZED}>:NORMALIZED>
          $<$<BOOL:${NO_X86}>:NO_X86>
          )
 
diff --git a/src/fs.c b/src/fs.c
index 21081dd..d00280c 100644
--- a/src/fs.c
+++ b/src/fs.c
@@ -25,7 +25,7 @@ int gk_fexists(char *fname)
   if (stat(fname, &status) == -1)
     return 0;
 
-  return S_ISREG(status.st_mode);
+  return S_ISREG(status.st_mode) | S_ISFIFO(status.st_mode);
 }
 
 
diff --git a/src/random.c b/src/random.c
index 3698614..476dfa9 100644
--- a/src/random.c
+++ b/src/random.c
@@ -23,7 +23,7 @@ GK_MKRANDOM(gk_idx, size_t, gk_idx_t)
 GK_MKRANDOM(gk_z,   size_t, ssize_t)
 GK_MKRANDOM(gk_zu,  size_t, size_t)
 
-
+#ifndef NORMALIZED
 
 /*************************************************************************/
 /*! GKlib's built in random number generator for portability across 
@@ -133,4 +133,4 @@ uint32_t gk_randint32(void)
 #endif
 }
 
-
+#endif /* NORMALIZED */
