diff --git a/Makefile b/Makefile
deleted file mode 100644
index 8fe67e7..0000000
--- a/Makefile
+++ /dev/null
@@ -1,101 +0,0 @@
-# Configuration options.
-i64        = not-set
-r64        = not-set
-gdb        = not-set
-assert     = not-set
-assert2    = not-set
-debug      = not-set
-gprof      = not-set
-valgrind   = not-set
-openmp     = not-set
-shared     = not-set
-cc         = not-set
-prefix     = ~/local
-gklib_path = ~/local
-
-
-# Basically proxies everything to the builddir cmake.
-
-BUILDDIR = build
-
-IDXWIDTH  = "\#define IDXTYPEWIDTH 32"
-REALWIDTH = "\#define REALTYPEWIDTH 32"
-
-# Process configuration options.
-CONFIG_FLAGS = -DCMAKE_VERBOSE_MAKEFILE=1
-ifneq ($(gklib_path), not-set)
-    CONFIG_FLAGS += -DGKLIB_PATH=$(abspath $(gklib_path))
-endif
-ifneq ($(prefix), not-set)
-    CONFIG_FLAGS += -DCMAKE_INSTALL_PREFIX=$(prefix)
-endif
-ifneq ($(i64), not-set)
-    IDXWIDTH  = "\#define IDXTYPEWIDTH 64"
-endif
-ifneq ($(r64), not-set)
-    REALWIDTH = "\#define REALTYPEWIDTH 64"
-endif
-ifneq ($(gdb), not-set)
-    CONFIG_FLAGS += -DGDB=$(gdb)
-endif
-ifneq ($(assert), not-set)
-    CONFIG_FLAGS += -DASSERT=$(assert)
-endif
-ifneq ($(assert2), not-set)
-    CONFIG_FLAGS += -DASSERT2=$(assert2)
-endif
-ifneq ($(debug), not-set)
-    CONFIG_FLAGS += -DDEBUG=$(debug)
-endif
-ifneq ($(gprof), not-set)
-    CONFIG_FLAGS += -DGPROF=$(gprof)
-endif
-ifneq ($(valgrind), not-set)
-    CONFIG_FLAGS += -DVALGRIND=$(valgrind)
-endif
-ifneq ($(openmp), not-set)
-    CONFIG_FLAGS += -DOPENMP=$(openmp)
-endif
-ifneq ($(shared), not-set)
-    CONFIG_FLAGS += -DSHARED=1
-endif
-ifneq ($(cc), not-set)
-    CONFIG_FLAGS += -DCMAKE_C_COMPILER=$(cc)
-endif
-
-VERNUM=5.1.0
-PKGNAME=metis-$(VERNUM)
-
-define run-config
-mkdir -p $(BUILDDIR)
-mkdir -p $(BUILDDIR)/xinclude
-echo $(IDXWIDTH) > $(BUILDDIR)/xinclude/metis.h
-echo $(REALWIDTH) >> $(BUILDDIR)/xinclude/metis.h
-cat include/metis.h >> $(BUILDDIR)/xinclude/metis.h
-cp include/CMakeLists.txt $(BUILDDIR)/xinclude
-cd $(BUILDDIR) && cmake $(CURDIR) $(CONFIG_FLAGS)
-endef
-
-all clean install:
-	@if [ ! -f $(BUILDDIR)/Makefile ]; then \
-		more BUILD.txt; \
-	else \
-	  	make -C $(BUILDDIR) $@ $(MAKEFLAGS); \
-	fi
-
-uninstall:
-	xargs rm < $(BUILDDIR)/install_manifest.txt
-
-config: distclean
-	$(run-config)
-
-distclean:
-	rm -rf $(BUILDDIR)
-
-remake:
-	find . -name CMakeLists.txt -exec touch {} ';'
-
-dist:
-	utils/mkdist.sh $(PKGNAME)
-
-.PHONY: config distclean all clean install uninstall remake dist
diff --git a/conf/gkbuild.cmake b/conf/gkbuild.cmake
index 591ece4..e849673 100644
--- a/conf/gkbuild.cmake
+++ b/conf/gkbuild.cmake
@@ -12,6 +12,7 @@ option(VALGRIND "add valgrind support" OFF)
 option(OPENMP "enable OpenMP support" OFF)
 option(PCRE "enable PCRE support" OFF)
 option(GKREGEX "enable GKREGEX support" OFF)
+option(NORMALIZED "enable normalization support" OFF)
 option(GKRAND "enable GKRAND support" OFF)
 
 # Add compiler flags.
@@ -106,6 +107,10 @@ if(GKREGEX)
   set(GK_COPTIONS "${GK_COPTIONS} -DUSE_GKREGEX")
 endif(GKREGEX)
 
+if(NORMALIZED)
+  set(GK_COPTIONS "${GK_COPTIONS} -DNORMALIZED")
+endif(NORMALIZED)
+
 if(GKRAND)
   set(GK_COPTIONS "${GK_COPTIONS} -DUSE_GKRAND")
 endif(GKRAND)
diff --git a/include/metis.h b/include/metis.h
index 5000d1b..2e6cae3 100644
--- a/include/metis.h
+++ b/include/metis.h
@@ -292,7 +292,8 @@ typedef enum {
   METIS_OPTION_NOOUTPUT,
   METIS_OPTION_BALANCE,
   METIS_OPTION_GTYPE,
-  METIS_OPTION_UBVEC
+  METIS_OPTION_UBVEC,
+  METIS_OPTION_OUTFILE
 } moptions_et;
 
 
diff --git a/libmetis/CMakeLists.txt b/libmetis/CMakeLists.txt
index fc6cec6..0c5f784 100644
--- a/libmetis/CMakeLists.txt
+++ b/libmetis/CMakeLists.txt
@@ -7,6 +7,8 @@ file(GLOB metis_sources *.c)
 # Build libmetis.
 add_library(metis ${METIS_LIBRARY_TYPE} ${metis_sources})
 
+target_link_libraries(metis functions)
+
 if(METIS_INSTALL)
   install(TARGETS metis
     LIBRARY DESTINATION lib
diff --git a/libmetis/defs.h b/libmetis/defs.h
index c18350f..50a7470 100644
--- a/libmetis/defs.h
+++ b/libmetis/defs.h
@@ -15,7 +15,12 @@
 #ifndef _LIBMETIS_DEFS_H_
 #define _LIBMETIS_DEFS_H_
 
+#ifdef NORMALIZED
+#define METISTITLE              "METIS NORMALIZED OUTPUT\n"
+#else
 #define METISTITLE              "METIS 5.2.1 Copyright 1998-22, Regents of the University of Minnesota\n"
+#endif
+
 #define MAXLINE			1280000
 
 #define LTERM			(void **) 0	/* List terminator for gk_free() */
diff --git a/libmetis/gklib.c b/libmetis/gklib.c
index 4e17eac..0b22780 100644
--- a/libmetis/gklib.c
+++ b/libmetis/gklib.c
@@ -75,12 +75,14 @@ void rsortd(size_t n, real_t *base)
 #undef r_gt
 }
 
+/*
 void ikvsorti(size_t n, ikv_t *base)
 {
 #define ikey_lt(a, b) ((a)->key < (b)->key)
   GK_MKQSORT(ikv_t, base, n, ikey_lt);
 #undef ikey_lt
 }
+*/
 
 /* Sorts based both on key and val */
 void ikvsortii(size_t n, ikv_t *base)
@@ -90,12 +92,14 @@ void ikvsortii(size_t n, ikv_t *base)
 #undef ikeyval_lt
 }
 
+/*
 void ikvsortd(size_t n, ikv_t *base)
 {
 #define ikey_gt(a, b) ((a)->key > (b)->key)
   GK_MKQSORT(ikv_t, base, n, ikey_gt);
 #undef ikey_gt
 }
+*/
 
 void rkvsorti(size_t n, rkv_t *base)
 {
diff --git a/libmetis/options.c b/libmetis/options.c
index 3f846cf..f5aab13 100644
--- a/libmetis/options.c
+++ b/libmetis/options.c
@@ -274,8 +274,13 @@ void PrintCtrl(ctrl_t *ctrl)
       if (i%modnum == 0)
         printf("\n     ");
       printf("%4"PRIDX"=[", i);
+#ifndef NORMALIZED
       for (j=0; j<ctrl->ncon; j++) 
         printf("%s%.2e", (j==0 ? "" : " "), (double)ctrl->tpwgts[i*ctrl->ncon+j]);
+#else
+      for (j=0; j<ctrl->ncon; j++) 
+        printf("%s%.5f", (j==0 ? "" : " "), (double)ctrl->tpwgts[i*ctrl->ncon+j]);
+#endif
       printf("]");
     }
     printf("\n");
diff --git a/libmetis/wspace.c b/libmetis/wspace.c
index 92853c3..878eae8 100644
--- a/libmetis/wspace.c
+++ b/libmetis/wspace.c
@@ -79,7 +79,11 @@ void AllocateRefinementWorkSpace(ctrl_t *ctrl, idx_t nbrpoolsize_max, idx_t nbrp
 /*************************************************************************/
 void FreeWorkSpace(ctrl_t *ctrl)
 {
+#ifndef NORMALIZED
   gk_mcoreDestroy(&ctrl->mcore, ctrl->dbglvl&METIS_DBG_INFO);
+#else
+  gk_mcoreDestroy(&ctrl->mcore, 0);
+#endif
 
   IFSET(ctrl->dbglvl, METIS_DBG_INFO,
       printf(" nbrpool statistics\n" 
diff --git a/programs/cmdline_gpmetis.c b/programs/cmdline_gpmetis.c
index 4260584..74f830d 100644
--- a/programs/cmdline_gpmetis.c
+++ b/programs/cmdline_gpmetis.c
@@ -45,6 +45,8 @@ static struct gk_option long_options[] = {
 
   {"dbglvl",         1,      0,      METIS_OPTION_DBGLVL},
 
+  {"outfile",        1,      0,      METIS_OPTION_OUTFILE},
+
   {"help",           0,      0,      METIS_OPTION_HELP},
   {0,                0,      0,      0}
 };
@@ -370,6 +372,10 @@ params_t *parse_cmdline(int argc, char *argv[])
         if (gk_optarg) params->dbglvl = (idx_t)atoi(gk_optarg);
         break;
 
+      case METIS_OPTION_OUTFILE:
+        if (gk_optarg) params->outfile = gk_strdup(gk_optarg);
+        break;
+
       case METIS_OPTION_HELP:
         for (i=0; strlen(helpstr[i]) > 0; i++)
           printf("%s\n", helpstr[i]);
diff --git a/programs/cmdline_mpmetis.c b/programs/cmdline_mpmetis.c
index 99c48ef..8bd85e7 100644
--- a/programs/cmdline_mpmetis.c
+++ b/programs/cmdline_mpmetis.c
@@ -37,6 +37,7 @@ static struct gk_option long_options[] = {
   {"seed",           1,      0,      METIS_OPTION_SEED},
 
   {"dbglvl",         1,      0,      METIS_OPTION_DBGLVL},
+  {"outfile",        1,      0,      METIS_OPTION_OUTFILE},
 
   {"help",           0,      0,      METIS_OPTION_HELP},
   {0,                0,      0,      0}
@@ -315,6 +316,10 @@ params_t *parse_cmdline(int argc, char *argv[])
         if (gk_optarg) params->dbglvl = (idx_t)atoi(gk_optarg);
         break;
 
+      case METIS_OPTION_OUTFILE:
+        if (gk_optarg) params->outfile = gk_strdup(gk_optarg);
+        break;
+
       case METIS_OPTION_HELP:
         for (i=0; strlen(helpstr[i]) > 0; i++)
           printf("%s\n", helpstr[i]);
diff --git a/programs/cmdline_ndmetis.c b/programs/cmdline_ndmetis.c
index a8f501c..0ddff2a 100644
--- a/programs/cmdline_ndmetis.c
+++ b/programs/cmdline_ndmetis.c
@@ -28,6 +28,8 @@ static struct gk_option long_options[] = {
   {"nseps",          1,      0,      METIS_OPTION_NSEPS},
   {"seed",           1,      0,      METIS_OPTION_SEED},
   {"dbglvl",         1,      0,      METIS_OPTION_DBGLVL},
+  // TODO: fixup WritePermutation
+  // {"outfile",        1,      0,      METIS_OPTION_OUTFILE},
   {"help",           0,      0,      METIS_OPTION_HELP},
   {0,                0,      0,      0}
 };
@@ -255,6 +257,10 @@ params_t *parse_cmdline(int argc, char *argv[])
         if (gk_optarg) params->dbglvl = (idx_t)atoi(gk_optarg);
         break;
 
+      case METIS_OPTION_OUTFILE:
+        if (gk_optarg) params->outfile = gk_strdup(gk_optarg);
+        break;
+
       case METIS_OPTION_HELP:
         for (i=0; strlen(helpstr[i]) > 0; i++)
           printf("%s\n", helpstr[i]);
diff --git a/programs/gpmetis.c b/programs/gpmetis.c
index 4b7e0b5..92e846e 100644
--- a/programs/gpmetis.c
+++ b/programs/gpmetis.c
@@ -126,7 +126,8 @@ int main(int argc, char *argv[])
     if (!params->nooutput) {
       /* Write the solution */
       gk_startcputimer(params->iotimer);
-      WritePartition(params->filename, part, graph->nvtxs, params->nparts); 
+      WritePartition(params->outfile ? params->outfile : params->filename,
+                     part, graph->nvtxs, params->nparts); 
       gk_stopcputimer(params->iotimer);
     }
 
@@ -232,6 +233,8 @@ void GPReportResults(params_t *params, graph_t *graph, idx_t *part, idx_t objval
 
   gk_stopcputimer(params->reporttimer);
 
+#ifndef NORMALIZED
+
   printf("\nTiming Information ----------------------------------------------------------\n");
   printf("  I/O:          \t\t %7.3"PRREAL" sec\n", gk_getcputimer(params->iotimer));
   printf("  Partitioning: \t\t %7.3"PRREAL" sec   (METIS time)\n", gk_getcputimer(params->parttimer));
@@ -246,6 +249,7 @@ void GPReportResults(params_t *params, graph_t *graph, idx_t *part, idx_t objval
     printf("  rusage.ru_maxrss:\t\t %7.3"PRREAL" MB\n", (real_t)(usage.ru_maxrss/(1024.0)));
   }
   printf("  proc/self/stat/VmPeak:\t %7.3"PRREAL" MB\n", (real_t)gk_GetProcVmPeak()/(1024.0*1024.0));
+#endif
 #endif
 
   printf("******************************************************************************\n");
diff --git a/programs/io.c b/programs/io.c
index 1d07fd6..4ac4d4f 100644
--- a/programs/io.c
+++ b/programs/io.c
@@ -428,9 +428,14 @@ void WritePartition(char *fname, idx_t *part, idx_t n, idx_t nparts)
   idx_t i;
   char filename[MAXLINE];
 
-  sprintf(filename, "%s.part.%"PRIDX, fname, nparts);
+  /* modify fname if ends in .graph, otherwise use as is */
+  if (strlen(fname) > strlen(".graph") 
+      && !strcmp(fname + (strlen(fname) - strlen(".graph")), ".graph")) {
+    snprintf(filename, MAXLINE, "%s.part.%"PRIDX, fname, nparts);
+    fname = filename;
+  }
 
-  fpout = gk_fopen(filename, "w", __func__);
+  fpout = gk_fopen(fname, "w", __func__);
 
   for (i=0; i<n; i++)
     fprintf(fpout,"%" PRIDX "\n", part[i]);
diff --git a/programs/m2gmetis.c b/programs/m2gmetis.c
index bb2f2b0..661fea2 100644
--- a/programs/m2gmetis.c
+++ b/programs/m2gmetis.c
@@ -128,6 +128,7 @@ void M2GReportResults(params_t *params, mesh_t *mesh, graph_t *graph)
 
   gk_stopcputimer(params->reporttimer);
 
+#ifndef NORMALIZED
 
   printf("\nTiming Information ----------------------------------------------------------\n");
   printf("  I/O:          \t\t %7.3"PRREAL" sec\n", gk_getcputimer(params->iotimer));
@@ -137,4 +138,6 @@ void M2GReportResults(params_t *params, mesh_t *mesh, graph_t *graph)
   printf("  Max memory used:\t\t %7.3"PRREAL" MB\n", (real_t)(params->maxmemory/(1024.0*1024.0)));
   printf("******************************************************************************\n");
 
+#endif
+
 }
diff --git a/programs/mpmetis.c b/programs/mpmetis.c
index 04a6a1b..de8d0a3 100644
--- a/programs/mpmetis.c
+++ b/programs/mpmetis.c
@@ -180,6 +180,7 @@ void MPReportResults(params_t *params, mesh_t *mesh, idx_t *epart, idx_t *npart,
 
   gk_stopcputimer(params->reporttimer);
 
+#ifndef NORMALIZED
 
   printf("\nTiming Information ----------------------------------------------------------\n");
   printf("  I/O:          \t\t %7.3"PRREAL" sec\n", gk_getcputimer(params->iotimer));
@@ -195,6 +196,7 @@ void MPReportResults(params_t *params, mesh_t *mesh, idx_t *epart, idx_t *npart,
     printf("  rusage.ru_maxrss:\t\t %7.3"PRREAL" MB\n", (real_t)(usage.ru_maxrss/(1024.0)));
   }
   printf("  proc/self/stat/VmPeak:\t %7.3"PRREAL" MB\n", (real_t)gk_GetProcVmPeak()/(1024.0*1024.0));
+#endif
 #endif
 
   printf("******************************************************************************\n");
diff --git a/programs/ndmetis.c b/programs/ndmetis.c
index 5991f0c..e1a1c8e 100644
--- a/programs/ndmetis.c
+++ b/programs/ndmetis.c
@@ -164,6 +164,7 @@ void NDReportResults(params_t *params, graph_t *graph, idx_t *perm,
 
   gk_stopcputimer(params->reporttimer);
 
+#ifndef NORMALIZED
 
   printf("\nTiming Information ----------------------------------------------------------\n");
   printf("  I/O:          \t\t %7.3"PRREAL" sec\n", gk_getcputimer(params->iotimer));
@@ -179,6 +180,7 @@ void NDReportResults(params_t *params, graph_t *graph, idx_t *perm,
     printf("  rusage.ru_maxrss:\t\t %7.3"PRREAL" MB\n", (real_t)(usage.ru_maxrss/(1024.0)));
   }
   printf("  proc/self/stat/VmPeak:\t %7.3"PRREAL" MB\n", (real_t)gk_GetProcVmPeak()/(1024.0*1024.0));
+#endif
 #endif
 
   printf("******************************************************************************\n");
