diff --git a/conf/gkbuild.cmake b/conf/gkbuild.cmake
index 591ece4..e849673 100644
--- a/conf/gkbuild.cmake
+++ b/conf/gkbuild.cmake
@@ -12,6 +12,7 @@ option(VALGRIND "add valgrind support" OFF)
 option(OPENMP "enable OpenMP support" OFF)
 option(PCRE "enable PCRE support" OFF)
 option(GKREGEX "enable GKREGEX support" OFF)
+option(NORMALIZED "enable normalization support" OFF)
 option(GKRAND "enable GKRAND support" OFF)
 
 # Add compiler flags.
@@ -106,6 +107,10 @@ if(GKREGEX)
   set(GK_COPTIONS "${GK_COPTIONS} -DUSE_GKREGEX")
 endif(GKREGEX)
 
+if(NORMALIZED)
+  set(GK_COPTIONS "${GK_COPTIONS} -DNORMALIZED")
+endif(NORMALIZED)
+
 if(GKRAND)
   set(GK_COPTIONS "${GK_COPTIONS} -DUSE_GKRAND")
 endif(GKRAND)
diff --git a/libmetis/CMakeLists.txt b/libmetis/CMakeLists.txt
index fc6cec6..0c5f784 100644
--- a/libmetis/CMakeLists.txt
+++ b/libmetis/CMakeLists.txt
@@ -7,6 +7,8 @@ file(GLOB metis_sources *.c)
 # Build libmetis.
 add_library(metis ${METIS_LIBRARY_TYPE} ${metis_sources})
 
+target_link_libraries(metis functions)
+
 if(METIS_INSTALL)
   install(TARGETS metis
     LIBRARY DESTINATION lib
diff --git a/libmetis/defs.h b/libmetis/defs.h
index c18350f..50a7470 100644
--- a/libmetis/defs.h
+++ b/libmetis/defs.h
@@ -15,7 +15,12 @@
 #ifndef _LIBMETIS_DEFS_H_
 #define _LIBMETIS_DEFS_H_
 
+#ifdef NORMALIZED
+#define METISTITLE              "METIS NORMALIZED OUTPUT\n"
+#else
 #define METISTITLE              "METIS 5.2.1 Copyright 1998-22, Regents of the University of Minnesota\n"
+#endif
+
 #define MAXLINE			1280000
 
 #define LTERM			(void **) 0	/* List terminator for gk_free() */
diff --git a/programs/gpmetis.c b/programs/gpmetis.c
index 4b7e0b5..bb9ef45 100644
--- a/programs/gpmetis.c
+++ b/programs/gpmetis.c
@@ -232,6 +232,8 @@ void GPReportResults(params_t *params, graph_t *graph, idx_t *part, idx_t objval
 
   gk_stopcputimer(params->reporttimer);
 
+#ifndef NORMALIZED
+
   printf("\nTiming Information ----------------------------------------------------------\n");
   printf("  I/O:          \t\t %7.3"PRREAL" sec\n", gk_getcputimer(params->iotimer));
   printf("  Partitioning: \t\t %7.3"PRREAL" sec   (METIS time)\n", gk_getcputimer(params->parttimer));
@@ -246,6 +248,7 @@ void GPReportResults(params_t *params, graph_t *graph, idx_t *part, idx_t objval
     printf("  rusage.ru_maxrss:\t\t %7.3"PRREAL" MB\n", (real_t)(usage.ru_maxrss/(1024.0)));
   }
   printf("  proc/self/stat/VmPeak:\t %7.3"PRREAL" MB\n", (real_t)gk_GetProcVmPeak()/(1024.0*1024.0));
+#endif
 #endif
 
   printf("******************************************************************************\n");
diff --git a/programs/m2gmetis.c b/programs/m2gmetis.c
index bb2f2b0..661fea2 100644
--- a/programs/m2gmetis.c
+++ b/programs/m2gmetis.c
@@ -128,6 +128,7 @@ void M2GReportResults(params_t *params, mesh_t *mesh, graph_t *graph)
 
   gk_stopcputimer(params->reporttimer);
 
+#ifndef NORMALIZED
 
   printf("\nTiming Information ----------------------------------------------------------\n");
   printf("  I/O:          \t\t %7.3"PRREAL" sec\n", gk_getcputimer(params->iotimer));
@@ -137,4 +138,6 @@ void M2GReportResults(params_t *params, mesh_t *mesh, graph_t *graph)
   printf("  Max memory used:\t\t %7.3"PRREAL" MB\n", (real_t)(params->maxmemory/(1024.0*1024.0)));
   printf("******************************************************************************\n");
 
+#endif
+
 }
diff --git a/programs/mpmetis.c b/programs/mpmetis.c
index 04a6a1b..de8d0a3 100644
--- a/programs/mpmetis.c
+++ b/programs/mpmetis.c
@@ -180,6 +180,7 @@ void MPReportResults(params_t *params, mesh_t *mesh, idx_t *epart, idx_t *npart,
 
   gk_stopcputimer(params->reporttimer);
 
+#ifndef NORMALIZED
 
   printf("\nTiming Information ----------------------------------------------------------\n");
   printf("  I/O:          \t\t %7.3"PRREAL" sec\n", gk_getcputimer(params->iotimer));
@@ -195,6 +196,7 @@ void MPReportResults(params_t *params, mesh_t *mesh, idx_t *epart, idx_t *npart,
     printf("  rusage.ru_maxrss:\t\t %7.3"PRREAL" MB\n", (real_t)(usage.ru_maxrss/(1024.0)));
   }
   printf("  proc/self/stat/VmPeak:\t %7.3"PRREAL" MB\n", (real_t)gk_GetProcVmPeak()/(1024.0*1024.0));
+#endif
 #endif
 
   printf("******************************************************************************\n");
diff --git a/programs/ndmetis.c b/programs/ndmetis.c
index 5991f0c..e1a1c8e 100644
--- a/programs/ndmetis.c
+++ b/programs/ndmetis.c
@@ -164,6 +164,7 @@ void NDReportResults(params_t *params, graph_t *graph, idx_t *perm,
 
   gk_stopcputimer(params->reporttimer);
 
+#ifndef NORMALIZED
 
   printf("\nTiming Information ----------------------------------------------------------\n");
   printf("  I/O:          \t\t %7.3"PRREAL" sec\n", gk_getcputimer(params->iotimer));
@@ -179,6 +180,7 @@ void NDReportResults(params_t *params, graph_t *graph, idx_t *perm,
     printf("  rusage.ru_maxrss:\t\t %7.3"PRREAL" MB\n", (real_t)(usage.ru_maxrss/(1024.0)));
   }
   printf("  proc/self/stat/VmPeak:\t %7.3"PRREAL" MB\n", (real_t)gk_GetProcVmPeak()/(1024.0*1024.0));
+#endif
 #endif
 
   printf("******************************************************************************\n");
